### DISCUSSION TITLE
umqtt send config info ,but cant recvice response

### DISCUSSION THREAD

**Discussion by BlackCattleProgrammer996** [micropython/micropython#13448] [ESP32]
Hello, MicroPython engineer!  I am currently developing applications using MicroPython language on a WIFI module ( ESP32-S3), and one of the functions is MQTT Cilent. When the WIFI module is used as an MQTT client (version 3.1.1) to connect to other MQTT servers (not Aliyun MQTT servers), the connection can be successful, and the publish/subscribe function is normal; As the picture shows:
![Figure 1](https://github.com/micropython/micropython/assets/122498020/d9bf5c3d-a7a9-48fe-8738-63071d8253cf)

However, when the Ali Cloud MQTT server is connected, the device prints a Response timeout (that is, no response data is received) as shown in Figure 2:
![Figure 2](https://github.com/micropython/micropython/assets/122498020/58e08568-6079-4bff-a530-2e15f8c77f25)

According to the MQTT protocol specification, the MQTT server should respond to data regardless of success or failure, rather than without data. From this phenomenon, it is suspected that the MQTT Connect request initiated by the device has not reached the MQTT server, and we hope that your engineer can help to check and analyze it. If the problem is the MQTT inventory used by the device, can you solve the problem by replacing other MicroPython?

For example, using umqtt.robust2: [https://github.com/fizista/micropython-umqtt.robust2/blob/master/src/umqtt/robust2.py](url)

Or use MicroPython-Mqtt: [https://github.com/peterhinch/micropython-mqtt](url)

**Comment by BlackCattleProgrammer996** [comment]
![Figure3](https://github.com/micropython/micropython/assets/122498020/d1d1035e-b393-4caa-b1bb-9f958f66cb99)

**Comment by felixdoerre** [comment]
Hi, can you maybe provide a network capture of the connection, to show if/where the micropython device is misbehaving? The log that you provided is consistent with an MQTT server that just closes the connection after username/password has been sent (which is also, what the error message in the line below says), and so far, I believe this is what happened. So it is not a timeout, the server actively closed the connection. 

> From this phenomenon, it is suspected that the MQTT Connect request initiated by the device has not reached the MQTT server.

I think this is really implausible. As the connection happens over a TCP connection, a message that has not reached the server is treated differently from a timeout.

Maybe the server is expecting an ssl-wrapped connection, and you are connecting in plain (or the other way round)? Maybe Username/Password is wrong and this is how the MQTT server that you are connecting to behaves?

**Reply by BlackCattleProgrammer996** [reply]
Thank you for your response. Later, I attempted to connect to the Alibaba Cloud MQTT server using the pahomqtt.c driver, and I was able to establish a successful connection. I don't know the differences between the two drivers. I provide you with the link to this driver (https://help.aliyun.com/zh/apsaramq-for-mqtt/developer-reference/sdk-reference-demo-projects-1?spm=5176.2020520104.0.0.19b33f1bfXoXHy), hoping to resolve the issue. Currently, my plan is to incorporate pahomqtt.c as a C language library for MicroPython and then invoke it, hoping that this approach will enable me to connect to the Alibaba Cloud MQTT server.
